serverless/resource_deployment.go

```diff
@@ -0,0 +1,256 @@
+package serverless
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"golang.org/x/mod/sumdb/dirhash"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/cloudformation"
+
+	yaml "gopkg.in/yaml.v2"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/customdiff"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+type serverlessConfig struct {
+	Service string
+}
+
+func getServiceName(configPath string) (string, error) {
+	configBytes, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		return "", err
+	}
+
+	config := serverlessConfig{}
+	err = yaml.Unmarshal(configBytes, &config)
+
+	return config.Service, err
+}
+
+func hashServerlessDir(configPath string, packagePath string) (string, error) {
+	absolutePackagePath := filepath.Join(filepath.Dir(configPath), packagePath)
+	zipPath := filepath.Join(absolutePackagePath, "sls-provider.zip")
+
+	zipHash, err := dirhash.HashZip(zipPath, dirhash.Hash1)
+	if err != nil {
+		return "", err
+	}
+
+	// https://github.com/golang/mod/blob/master/sumdb/dirhash/hash.go#L75
+	osOpen := func(name string) (io.ReadCloser, error) {
+		return os.Open(name)
+	}
+
+	configHash, err := dirhash.Hash1([]string{configPath}, osOpen)
+	if err != nil {
+		return "", err
+	}
+
+	return strings.Join([]string{zipHash, configHash}, "-"), nil
+}
+
+type serverlessParams struct {
+	command           string
+	serverlessBinPath string
+	configPath        string
+	packageDir        string
+	stage             string
+	args              []interface{}
+}
+
+func runServerless(params *serverlessParams) error {
+	stringArgs := make([]string, len(params.args))
+	for i, v := range stringArgs {
+		stringArgs[i] = fmt.Sprint(v)
+	}
+
+	requiredArgs := []string{
+		params.command,
+		"-s",
+		params.stage,
+	}
+
+	if params.command == "deploy" || params.command == "package" {
+		requiredArgs = append(requiredArgs, "-p", params.packageDir)
+	}
+
+	stringArgs = append(
+		requiredArgs,
+		stringArgs...,
+	)
+	cmd := exec.Command(params.serverlessBinPath, stringArgs...)
+	cmd.Dir = filepath.Dir(params.configPath)
+
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("%v\n%w", string(output), err)
+	}
+
+	return nil
+}
+
+func resourceDeployment() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceDeploymentCreate,
+		Read:   resourceDeploymentRead,
+		Update: resourceDeploymentUpdate,
+		Delete: resourceDeploymentDelete,
+
+		Schema: map[string]*schema.Schema{
+			"config_path": &schema.Schema{
```

Comment:
Serverless does this as automatic config inference: https://github.com/serverless/serverless/blob/1cad7b88fb92123e94ab2f5eae2e4f06ef7b3001/lib/utils/getServerlessConfigFile.js#L17-L39

In particular choosing config files based on this implied order:

```js
  const yamlPath = path.join(servicePath, 'serverless.yaml');
  const ymlPath = path.join(servicePath, 'serverless.yml');
  const jsonPath = path.join(servicePath, 'serverless.json');
  const jsPath = path.join(servicePath, 'serverless.js');
```

We could potentially offer a default option if not set:
- (TOTALLY OPTIONAL) Take `servicePath` arguments like SLS does.
- Do the same default config file order, which we should do for correctness if allowing a default.

... or require absolute config specification like I think we have now which is fine. But maybe document the above scenario just a bit?

Reply:
This actually exposes another problem–we parse the YAML and grab the service name from there, but not all Serverless configs are YAML! Will have a think on this one.

Reply:
Aha! I can use ` sls print --path service --format text`